File: algospot/시계맞추기/clocksync.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const int INF = 9999, SWITCH = 10, CLOCKS = 16;

const char linked[SWITCH][CLOCKS + 1] = {
	"xxx.............",
	"...x...x.x.x....",
	"....x.....x...xx",
	"x...xxxx........",
	"......xxx.x.x...",
	"x.x...........xx",
	"...x..........xx",
	"....xx.x......xx",
	".xxxxx..........",
	"...xxx...x...x.."
};

bool areAlined(const vector<int>& clocks){
	for (int i = 0; i < CLOCKS; i++){
		if (clocks[i] != 12) return false;
	}
	return true;
}

void push(vector<int>& clocks, int swtch){
	for(int i = 0 ;i < CLOCKS; i++){
		if (linked[swtch][i] == 'x'){
			clocks[i] += 3;
			if (clocks[i] == 15) clocks[i] = 3;
		}
	}
}

int solve(vector<int>& clocks, int swtch){
	if (swtch == SWITCH) return areAlined(clocks) ? 0 : INF;
	int ret = INF;
	for (int i = 0; i < 4; i++){
		ret = min(ret, i + solve(clocks, swtch + 1));
		push(clocks, swtch);
	}
	return ret;
}

int main(){
	int C;
	cin >> C;
	for (int i = 0; i < C; i++){
		vector<int> clocks;
		int a;
		for (int j = 0; j < CLOCKS; j++){
			cin >> a;
			clocks.push_back(a);
		}
		int ans = solve(clocks,0);
		if (ans == INF) cout << -1 << endl;
		else cout << ans << endl;
		clocks.clear();
	}
}

File: algospot/시계맞추기/clocksync.py
C = int(input())

switch = [[0,1,2], [3, 7, 9, 11], 
[4, 10, 14, 15], [0, 4, 5, 6, 7], 
[6, 7, 8, 10, 12], [0, 2, 14, 15],
[3, 14, 15],[4, 5, 7, 14, 15],
[1, 2, 3, 4, 5],[3, 4, 5, 9, 13]]

INF = 1000

def push(clocks, switchIdx):
	for i in switch[switchIdx]:
		clocks[i] += 3
		if clocks[i] >= 15:
			clocks[i] = 3

def checkClcock(clocks):
	for hour in clocks:
		if (hour != 12):
			return False
	return True

def solve(clocks, switchIdx):
	if switchIdx == 10:
		if checkClcock(clocks):
			return 0
		return INF
	minCount = INF
	for i in range(4):
		minCount = min(i + solve(clocks, switchIdx + 1), minCount)
		push(clocks, switchIdx)

	return minCount

for _ in range(C):
	clocks = list(map(int, input().split()))
	count = solve(clocks, 0)
	print(-1 if count == INF else count)
